### PR TITLE
UI: PageLayout component

### PR DESCRIPTION
**Pull request by kpyszkowski**
Ref: https://github.com/threshold-network/bitcoin-on-base/issues/2
~Depends on: https://github.com/threshold-network/bitcoin-on-base/pull/12~

This pull request introduces `PageLayout` component. I decided to rewrite the component entirely. The major feature is render slots to conditionally render content in specific areas regarding to designs.
<details>
<summary>Example combinations of prop values presenting possibilities:</summary>

- `renderTop` with value, `renderLeft/Right` equals `null`, `backgroundVariant` equals `'primary'`:
localhost_3000_tBTC_mint
- `renderTop` with value, `renderLeft` equals `undefined`,  `renderRight` equals `null`, `backgroundVariant` equals `'primary'`:
localhost_3000_tBTC_mint (1)
- `renderTop` equals `undefined`, `renderLeft` with value,  `renderRight` equals `undefined`, `backgroundVariant` equals `'primary'`:
localhost_3000_tBTC_mint (2)
- `renderTop` equals `undefined`, `renderLeft` equals `undefined`,  `renderRight` with value, `backgroundVariant` equals `'secondary'`:
localhost_3000_tBTC_mint (3)

</details>

Render slots are wrapped with padding container following design appearance, children is not wrapped with any container (padding) since it's appears to be variable in various views.

https://github.com/threshold-network/bitcoin-on-base/assets/11503879/deccabfb-3f6e-4f2f-a07a-76a162635e1b

Note: The TODO regarding Balance component (#17) isn't a blocker. It's just a placeholder value indicating to put the component there when the view is composed in separate PR.

### CODE REVIEW

File: src/pages/PageLayout.tsx
Comment (michalsmiarowski) [inline src/pages/PageLayout.tsx:79]: I'm not sure what was the point to refactor the `PageLayout` component tbh 🤔  I also don't think implementing it like this is a good idea.

The `PageLayout` should be reusable for every page and it should contain only necessary and very basic thing, that applies for every page. This is why we've only had `Container` component there as it's common for all the pages. The actual implementation assumes that the page will contain the maximum of 3 column.

The second thing is, that this implementation is not responsive at all. Decreasing the width of the screen decreases the width of the 'Outlet' component instead of reducing the margins on the sides.

image

Next thing, putting slots to render as props is also a bad idea imo. The `PageLayout` should be more "fluid" and we shouldn't assume that this will be reusable for all pages. For the `Minting` flow we should create a separate layout and name it differently. I think it will be anough if we create one horizontal stack (that is only visible when user is connected) and three rows. We can also use Grid for that. For the Bridge main page we should implement a separate layout that only contains two columns that are responsive. I believe the Bridge main page is not implemented in this PR so let's focus on minting layout first.

File: src/pages/tBTC/Bridge/index.tsx
Comment (michalsmiarowski) [inline src/pages/tBTC/Bridge/index.tsx:37]: What I would do here is I would create two separate layouts. One for `MintingProcess` and one for `Bridge` main page. They should have different URLs IMO:
- tbtc/bridge
- tbtc/bridge/mint

I think the main page for Bridge is not implemented in this PR so for now we could just create a MintLayout.

File: src/pages/PageLayout.tsx
Comment (michalsmiarowski) [inline src/pages/PageLayout.tsx:114]: Unnecessary double exclamation

File: src/pages/PageLayout.tsx
Comment (michalsmiarowski) [inline src/pages/PageLayout.tsx:118]: Same as above

File: src/pages/tBTC/index.tsx
Comment (michalsmiarowski) [inline src/pages/tBTC/index.tsx:10]: Let's destructure the `title` prop in the component parameters:

```
const MainTBTCPage: PageComponent = ({ title = "tBTC" }) => {
(...)
```

**Review by michalsmiarowski — CHANGES_REQUESTED** [review]
Let some comments to look at before the merge

File: src/pages/PageLayout.tsx
Comment (kpyszkowski) [reply src/pages/PageLayout.tsx:80]: > The `PageLayout` should be reusable for every page 

Actually it is. It covers every scenario presented in designs and can be freely used.

> This is why we've only had `Container` component

The `PageLayout` now acts as this component. My objective was to get rid of "pseudo" reusable components like `<Something>Layout` (+ confusing wrapper component somewhere in between) in favor of truely reusable, multipurpose `PageLayout` component.

I have implemented responsive styles. It behaves (almost) like Threshold Dashboard. Ref commit: 9a864bd2cdfc53d6095a07f97f6bba0fff5d2177

Now `Header` component from #13 might need some fine tuning regarding to alignment between menu and content but I'm not sure how it should behave. I'd leave as is for now. What do you think?

File: src/pages/tBTC/Bridge/index.tsx
Comment (kpyszkowski) [reply src/pages/tBTC/Bridge/index.tsx:37]: It's only the example how we could conditionally render sidebars.
As I explained [here](https://github.com/threshold-network/bitcoin-on-base/pull/15#discussion_r1466426079) it's pointless to create separate layouts, just consume `PageLayout` and control it via props.

File: src/pages/PageLayout.tsx
Comment (kpyszkowski) [reply src/pages/PageLayout.tsx]: Ref commit: 0258d93a

File: src/pages/PageLayout.tsx
Comment (kpyszkowski) [reply src/pages/PageLayout.tsx]: Ref commit: 0258d93a

File: src/App.tsx
Comment (kpyszkowski) [reply src/App.tsx:103]: Ref commit: 14976fcd
Additionally I suppose the `Navbar` component can be removed too. Ref commit: e774be02 
Let me know if you agree. I can revert it.

File: src/pages/tBTC/index.tsx
Comment (kpyszkowski) [reply src/pages/tBTC/index.tsx:10]: Ref commit: ea6140e3

File: src/pages/PageLayout.tsx
Comment (michalsmiarowski) [inline src/pages/PageLayout.tsx:15]: Can you explain how did you come up with `CONTENT_MAX_WIDTH` value ("89.25rem" // 1428px) exactly? 🤔

File: src/pages/PageLayout.tsx
Comment (michalsmiarowski) [inline src/pages/PageLayout.tsx:139]: I think we should add some padding to this box by default.

image

File: src/pages/PageLayout.tsx
Comment (michalsmiarowski) [inline src/pages/PageLayout.tsx:1]: Non-blocking but good to have:

Overall it would be great to use chakra-ui theme feature and create a separate PageLayout theme with the styles used here. The `CONTENT_MAX_WIDTH` and `BORDER` constatns could also be move there.

**Review by michalsmiarowski — CHANGES_REQUESTED** [review]
Left some more comments to look at before the merge

File: src/pages/PageLayout.tsx
Comment (michalsmiarowski) [reply src/pages/PageLayout.tsx:82]: Regarding the implementation, let's move the discussion here: https://github.com/threshold-network/bitcoin-on-base/pull/15#discussion_r1467402522

Regarding `Header` component and it's behaviour - I think we can leave it for now. We can improve it later if needed when we know more about how it should behave.

File: src/components/Navbar/NavbarComponent.tsx
Comment (michalsmiarowski) [inline src/components/Navbar/NavbarComponent.tsx:18]: Let's remove those two components.

File: src/components/Navbar/HamburgerButton.tsx
Comment (michalsmiarowski) [inline src/components/Navbar/HamburgerButton.tsx:4]: I think `useSidebar` can be remove too

File: src/components/Sidebar/BrandIcon.tsx
Comment (michalsmiarowski) [inline src/components/Sidebar/BrandIcon.tsx:7]: Let's remove those icons

File: src/components/Sidebar/SidebarFooter.tsx
Comment (michalsmiarowski) [inline src/components/Sidebar/SidebarFooter.tsx:46]: For now, let's remove `thresholdGithub` from `ExternalHref`. We can add it later if needed. It point to the wrong repo anyway.

**Review by michalsmiarowski — CHANGES_REQUESTED** [review]
Left few more comments related to `Sidebar` and `Navbar` removal.

File: src/App.tsx
Comment (michalsmiarowski) [reply src/App.tsx]: I agree. Just left few comments regarding that: https://github.com/threshold-network/bitcoin-on-base/pull/15#pullrequestreview-1845557097

File: src/pages/tBTC/HowItWorks/index.tsx
Comment (kpyszkowski) [reply src/pages/tBTC/HowItWorks/index.tsx:13]: It's intentional. It was designed before major refactor of the designs (v2 -> BoB/dApp)
image
The border disappears when the value is undefined reflecting designs conceptions.

File: src/pages/PageLayout.tsx
Comment (kpyszkowski) [reply src/pages/PageLayout.tsx:139]: There is a scenario with minting success page where the background element will be missplaced if the padding will be applied here.
I will introduce additional prop to conditonally render this padding around children.
image

File: src/pages/PageLayout.tsx
Comment (kpyszkowski) [reply src/pages/PageLayout.tsx:15]: From Figma. Measured distance between most outer vertical borders.
image

File: src/pages/tBTC/Bridge/index.tsx
Comment (kpyszkowski) [reply src/pages/tBTC/Bridge/index.tsx:50]: In this particular case it's pointless to leverage compound components pattern. 
The major advantage of this pattern is a **loose structure** with **communication between children**. We don't want loose structure in Layout, we know strictly what is the expected structure. Moreover we don't need communication. We just want to render specific content in specific area.

The example you provided does not present compound components pattern. Components don't have any shared state - they "don't know" about each other. They are just imported as an object so you can use dot notation while referring. Compound components share the same state (usually context) and are aware of state changes managed by the parent node (it's possible to make it bi-directional too).

[Render slots](https://daveceddia.com/pluggable-slots-in-react-components/) is the proper architectural pattern to use in this case.

> I strongly believe that passing components through props can get cumbersome

I don't understand. Render props are typed to accept only `ReactNode` exactly the same as children which in fact is just specific prop you define by passing value between opening and closing tags. 
```jsx
<Component>
  <div>content</div>
</Component>
```
is exactly the same as
```jsx
<Component children={<div>content</div>} />
```
It's impossible to mess it up, especially when it's strongly typed. I also took care of documentation and described each prop and component to prevent confusions when consuming component.

To sum up:
- compound components are suitable when you want loose structure and communication with shared state
- render slots are suitable when "children" is not enough - when you need to define multiple render areas to be present in specific places, you can treat is as "alternative children".

Here is [example of compound components pattern](https://github.com/kpyszkowski/Nardo-Primitives/blob/feature/tabs/src/components/Tabs/Tabs.tsx) I have created some time ago. It's not ideal buy presents key concepts of the pattern.

File: src/pages/PageLayout.tsx
Comment (kpyszkowski) [reply src/pages/PageLayout.tsx]: Ref commit: a6428b9c

File: src/pages/PageLayout.tsx
Comment (kpyszkowski) [reply src/pages/PageLayout.tsx]: Ref commit: 6654f51bf2

File: src/pages/PageLayout.tsx
Comment (kpyszkowski) [reply src/pages/PageLayout.tsx:1]: Ref commit: f484c030

File: src/components/Navbar/NavbarComponent.tsx
Comment (kpyszkowski) [reply src/components/Navbar/NavbarComponent.tsx:18]: Ref commit: eb2a53a2

File: src/components/Navbar/HamburgerButton.tsx
Comment (kpyszkowski) [reply src/components/Navbar/HamburgerButton.tsx:4]: Ref commit: eb2a53a2

File: src/components/Sidebar/SidebarFooter.tsx
Comment (kpyszkowski) [reply src/components/Sidebar/SidebarFooter.tsx:46]: Ref commit: eb2a53a2

File: src/pages/PageLayout.tsx
Comment (michalsmiarowski) [reply src/pages/PageLayout.tsx:1]: This is not what I intended. What I mean is to incorporate all (or most) styles into a separate PageLayout theme, similar to this:

```
export const PageLayout = {
  parts: [
    "pageLayoutContainer",
    "pageLayoutTopSection",
    "pageLayoutLeftSection",
    "pageLayoutRightSection",
    "pageLayoutMainSection",
  ],
  baseStyle: (props: any) => {
    return {
      pageLayoutContainer: {
        { paste style here}
      },
      pageLayoutTopSection: {
        { paste style here}
      },
      pageLayoutLeftSection: {
        { paste style here}
      },
      pageLayoutRightSection: {
        { paste style here}
      },
      pageLayoutMainSection: {
        { paste style here}
      },
    }
  },
  variants: {
    primary: {
      pageLayoutContainer: {
        { primary background styles here}
      },
    },
    secondary: {
      pageLayoutContainer: {
        { secondary background styles here}
      },
    }
  }
}

```

And then assign it in the component like:

```
const styles = useMultiStyleConfig("PageLayout", props)
(...)
  return (
    <Flex
      sx={styles.pageLayoutContainer}
      {...restProps}
    >
 (...etc...)
```

However, I haven't tested this myself. We should utilize the features that Chakra provides, as it will simplify the `PageLayout` component. For instance, by using variants, we can eliminate checks like `variant === "primary"`.

I suggest we handle this in a separate PR to not block this one.

File: src/pages/tBTC/Bridge/index.tsx
Comment (michalsmiarowski) [reply src/pages/tBTC/Bridge/index.tsx:50]: I just feel like the approach I proposed is a little bit more idiomatic and more natural to React in general. But we can have it your way too, I don't mind. Both solutions are pretty similar to each other.

File: src/hooks/useSidebar.ts
Comment (kpyszkowski) [reply src/hooks/useSidebar.ts:2]: Ref commit: 8804099b

File: src/hooks/useSidebar.ts
Comment (kpyszkowski) [reply src/hooks/useSidebar.ts:10]: Ref commit: 8804099b

**Review by michalsmiarowski — APPROVED** [review]
LGTM

**Pull request merged**
